woocommerce/woocommerce-android
# Jetpack sign in - Step 1: Fetch Jetpack e-mail address for flowing through Magic Links

Partially fixes #1482. This PR completes the first step of the Jetpack sign in with self credentials as defined in the master thread. 

### Changes
- ~~Revert existing logic of validating site url from `FETCH_CONNECT_SITE_INFO` endpoint and use the `FETCH_WPCOM_SITE_BY_URL` endpoint and initiate the discovery process.~~ Since using the discovery endpoint to validate the site would interfere with the existing logic of logging in using wp.com credentials, I've moved the discovery process to initiate only when the user enters their username + password in the `LoginUsernamePasswordFragment`.
- Added a new button to `LoginEmailFragment`: `Login via site credentials`.
- Design changes:
  - Added label to the top of the `LoginUsernamePasswordFragment`.
  - Change hint text of `username` in `LoginUsernamePasswordFragment`.
- When site credentials are clicked:
  - Initialise the discovery process which will let us know if the XMLRPC endpoint is supported by the site. **Note that the error handling during the discovery process will take place in another PR.**
   - If jetpack is installed/active/connected: call the `wp.getOptions` endpoint to fetch the `jetpack_user_email` and store in `SiteModel` and redirect to the request magic link screen.
   - If jetpack NOT installed/active/connected: Redirect to a new fragment called `LoginNoJetpackFragment`. 


### Notes:
- This PR handles only the redirection to the magic link login. The design changes required in the magic link login will take place in another PR.
- Since this PR adds logic to initiate the discovery process, there are new error messages and error scenarios that need to be incorporated. This will likely take place in another PR.
- ~Since this PR removes the logic to validate site address using the `FETCH_CONNECT_SITE_INFO` endpoint, there are some events that will likely be invalidated once this feature is merged.~
   - `LOGIN_CONNECTED_SITE_INFO_FAILED`
   - `LOGIN_CONNECTED_SITE_INFO_SUCCEEDED` 
- The are TODO's added to the PR to add events to the button clicks. This will take place in another PR.
- Note that this PR is to be merged to a feature branch. 


### Deviations from design:
- The username hint in `LoginUsernamePasswordFragment` is set to just `Username` rather than just `Username or Email` because there is no option to pass email address via the API, just the username.

### Behaviour
(LEFT: Login with self hosted site with Jetpack . RIGHT: Login with self hosted site without Jetpack)
<img width="300" src="https://user-images.githubusercontent.com/22608780/67195788-3c9f2a80-f417-11e9-97ba-7fb73e6e826a.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/67195794-3f018480-f417-11e9-8897-8afac854b4a2.gif">

##### Entering a non existent site:
<img width="300" src="https://user-images.githubusercontent.com/22608780/67195997-a4ee0c00-f417-11e9-9e75-c3da2669b6ad.gif">

#### Testing
- Verify that existing flow works as expected.
- Enter a non existent site and verify that an error message is displayed.
- Enter a non jetpack site and verify that Jetpack required screen is displayed.
  - Click on `Sign in` at the bottom of the screen.
  - Verify that the Email screen is displayed. Click on `Login with site credentials`.
  - Enter username + password and click on `Next`.
  - Verify that you are redirected to Jetpack needed screen.
  - Install, activate and connect Jetpack on your test site while in the app.
  - Click on `Refresh the app` link. Notice that you are redirected to the Site credentials screen.
  - Click on `Next` and verify that you are redirected to the Magic link sign in screen.
- Enter a jetpack site and verify that you are redirect to the email screen.
  - Click on `Login with site credentials`.
  - Enter username + password and click on `Next` and verify that you are redirected to the Magic link sign in screen.
  

Update release notes:
